vault/ui_custom_message_test.go

```diff
@@ -0,0 +1,611 @@
+	assert.Error(t, err)
+	assert.Nil(t, result)
+}
+
+// TestCreateCustomMessage verifies that the (*UIConfig).CreateCustomMessage
+// works as expected in all circumstances. This includes errors in the
+// underlying logical.Storage, reaching the maximum count of custom messages.
+func TestCreateCustomMessage(t *testing.T) {
+	testUIConfig := &UIConfig{
+		barrierStorage: &logical.InmemStorage{},
+	}
+
+	testUIConfig.nsBarrierView = NewBarrierView(testUIConfig.barrierStorage, "namespaces/")
+
+	entry := UICustomMessageEntry{
+		Title:         "title",
+		Message:       "message",
+		Authenticated: true,
+		MessageType:   "modal",
+		StartTime:     time.Now().Add(-1 * time.Hour),
+		EndTime:       time.Now().Add(time.Hour),
+	}
+
+	// Create a custom message
+	result, err := testUIConfig.CreateCustomMessage(context.Background(), entry)
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+	assert.NotEmpty(t, result.Id)
+	assert.True(t, result.active)
+
+	// Create more custom message to have the maximum number of custom messages.
+	for i := 0; i < MaximumCustomMessageCount-1; i++ {
+		result, err = testUIConfig.CreateCustomMessage(context.Background(), entry)
+		assert.NoError(t, err)
+		assert.NotNil(t, result)
+	}
+
+	// We should now have the maximum number of custom messages in storage,
+	// try to create one more to get an error.
+	result, err = testUIConfig.CreateCustomMessage(context.Background(), entry)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "maximum")
+	assert.Nil(t, result)
+
+	// Make sure that the time validation check occurs before the maximum
+	// custom messages check (by not only looking for an error, but checking
+	// that it's the failure error for the validateStartAndEndTimes function).
+	entryWithInvalidTimes := entry
+	entryWithInvalidTimes.StartTime = entry.EndTime
+	entryWithInvalidTimes.EndTime = entry.StartTime
+	result, err = testUIConfig.CreateCustomMessage(context.Background(), entryWithInvalidTimes)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "occur before")
+	assert.Nil(t, result)
+
+	// Make sure that a custom message can be still be created in a different
+	// namespace.
+	ns := &namespace.Namespace{
+		ID:   "abc",
+		Path: "def/",
+	}
+	result, err = testUIConfig.CreateCustomMessage(namespace.ContextWithNamespace(context.Background(), ns), entry)
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+	assert.NotEmpty(t, result.Id)
+
+	// Check that error is returned when an error occurs in the underlying
+	// logical.Storage.
+	testUIConfig.barrierStorage = &testingStorage{
+		listFails: true,
+	}
+
+	result, err = testUIConfig.CreateCustomMessage(context.Background(), entry)
+	assert.Error(t, err)
+	assert.Nil(t, result)
+
+	testUIConfig.barrierStorage = &testingStorage{
+		putFails: true,
+	}
+
+	result, err = testUIConfig.CreateCustomMessage(context.Background(), entry)
+	assert.Error(t, err)
+	assert.Nil(t, result)
+}
+
+// TestDeleteCustomMessage verifies
```

Comment:
![I think this godoc needs to be finished.](https://media.tenor.com/yqJt4RkqaxAAAAAC/keep-your-secrets-frodo.gif)